### DISCUSSION TITLE
How to properly deserialize

### DISCUSSION THREAD

**Discussion by kiwidoggie** [protobuf-c/protobuf-c#605] [Q&A] [answered]
Hello, I was looking through the example usage of protobuf c. I have an incoming buffer of X bytes as an example, but when calling unpack on the X bytes, it returns NULL. I noticed in the examples you use a static buffer (larger than the message bytes), and was wondering is this always needed? Or is there a way to get the unpacked size (similar to how we get the packed size) thanks :D

(not a bug, more of question)

**Comment by ubx-luis** [comment]
I have tried deserialize simple messages (no submessages) and it works fines... It didnt needed a bigger buffer than what its needed to store it (no operations on adjacent memory was done).
The examples seems to assume that only one protobuf is send at the time and there is nothing else to read there, in which case MAX_MSG_SIZE is just big enough to not cause any problem with the examples... Is very likely that you will have to use some sort of protocol that indicates when a protobuf is send and the length of it.

I'm having issues deserializing messages with multiple submessages inside, but seems unrelated to the size of the buffer.

**Comment by lipnitsk** [comment, marked as answer]
protobuf-c uses dynamic memory allocation when unpacking messages, so there is no way to know exactly how much memory will be used. But you could potentially use a memory pool with a custom allocator, sized for the worst case scenario based on your use case.
